### DISCUSSION TITLE
Head/Tail of a csv file in table format

### DISCUSSION THREAD

**Discussion by StatPal** [nushell/nushell#4026] [Q&A] [answered]
If I open a large CSV file, by `open filename.csv` or even `open --raw filename.csv`, it shows a nice table-like format. 

Now, suppose I want to see only the first 20 lines, or more importantly the last 10 lines, possibly through `head` or `tail` commands. How do I do that? I have tried `head`/`tail` commands - but they look like normal `head`/`tail` are not that good-looking. Is there any way to tackle this? 

Thanks.

**Comment by sophiajt** [comment, marked as answer]
you could try `open filename.csv | first 10` or `open filename.csv | last 10`

**Reply by StatPal** [reply]
Thank you very much. That works like a charm. :)

**Reply by pedroangelini** [reply]
what if the file is very large? Is there a way to avoid that nushell open the whole file? I'm talking 2 GB file here --- fits in memory, but takes forever

**Reply by fdncred** [reply]
My first guess would be to try something like `open big.csv | into binary | bytes at 12000.. | decode utf8`. I'm not sure this would work but it's where i'd start. This should read the file starting at whatever bytes you specify until the end and decode spits it out as utf8.

**Reply by ghost** [reply]
its not pure nushell, but head and tail are still available if its installed in the system -> `head -n10 big.csv | from csv` should work (with `--noheaders` if you use tail since the headers will obviously be missing)

**Comment by t0w2** [comment]
Is there a way to show the first and last at the same time? something like this `open abc.csv | { first 5; last 5; }`

**Reply by fdncred** [reply]
I don't think there's a built-in command to print the first and last. You could write a custom command that does it or something like this.
```nushell
ls | first) | append (ls | last)
```

**Comment by NotTheDr01ds** [comment]
@t0w2 Sorry for the delay (I always forget Discussions are a thing here).  

List spreading syntax can also be used here. I tend to prefer this to `append` in cases like this because it's easier to see the visual representation of how you want the data "placed" in the list:

```nu
open Cargo.toml --raw | lines
| [
    ...($in | first 5)
    '....'
    ...($in | last 5)
  ]

# =>
╭────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────╮
│              0 │ [package]                                                                                        │
│              1 │ authors = ["The Nushell Project Developers"]                                                     │
│              2 │ build = "scripts/build.rs"                                                                       │
│              3 │ default-run = "nu"                                                                               │
│              4 │ description = "A new type of shell"                                                              │
│              5 │ ....                                                                                             │
│              6 │ # Run all benchmarks with `cargo bench`                                                          │
│              7 │ # Run individual benchmarks like `cargo bench -- <regex>` e.g. `cargo bench -- parse`            │
│              8 │ [[bench]]                                                                                        │
│              9 │ name = "benchmarks"                                                                              │
│             10 │ harness = false                                                                                  │
╰────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────╯
```

The `...($in | first 5)` expands the first 5 items into individual elements that can be placed in a new list, etc.

Keep in mind that if the file is less than 10 lines, you'll get duplicates.
